In [1]:
import numpy as np
from qiskit import QuantumCircuit, Aer, transpile
from qiskit.visualization import plot_bloch_multivector
from qiskit.quantum_info import Statevector
from IPython.display import display
from qiskit.tools.visualization import plot_histogram

sim = Aer.get_backend("aer_simulator")

# Q11 GHZ状態(3量子ビットのもつれ)について、前問と同様に確認せよ。

## 解答

In [2]:
qc = QuantumCircuit(3, 3)
qc.h(qubit=0)
qc.cx(control_qubit=0, target_qubit=1)
qc.cx(control_qubit=1, target_qubit=2)
state_vec = Statevector(data=qc).to_dict()

for bitstring, wave_func in state_vec.items():
    prob = wave_func.real**2 + wave_func.imag**2
    print(f"Bitstring {bitstring}: "
          f"Expected observation prob. = {np.round(prob, decimals=8)}")

qc.measure(qubit=[0, 1, 2], cbit=[0, 1, 2])
shots = 100000
counts = sim.run(circuits=qc, shots=shots).result().get_counts()
for bitstring, count in counts.items():
    print(f"Bitstring {bitstring}: "
          f"Actual observation prob. = {np.round(count/shots, decimals=2)}")

Bitstring 000: Expected observation prob. = 0.5
Bitstring 111: Expected observation prob. = 0.5
Bitstring 111: Actual observation prob. = 0.5
Bitstring 000: Actual observation prob. = 0.5


## 解説
3量子ビットのGHZ状態は、3量子ビットの量子回路において、アダマールゲートと制御Xゲートを実行することで得られる。初期状態が$|000\rangle$のケースを具体的に計算してみる。まずアダマールゲートを0番目の量子ビットに作用させると
$$
|000\rangle \to \frac{1}{\sqrt{2}}(|000\rangle+|001\rangle)
$$
が得られる。次に、0番目の量子ビットを制御ビットとする制御Xゲートを1番目の量子ビットに作用させると
$$
\frac{1}{\sqrt{2}}(|000\rangle+|001\rangle) \to \frac{1}{\sqrt{2}}(|000\rangle+|011\rangle)
$$
が得られる。最後に、1番目の量子ビットを制御ビットとする制御Xゲートを2番目の量子ビットに作用させると
$$
\frac{1}{\sqrt{2}}(|000\rangle+|011\rangle) \to \frac{1}{\sqrt{2}}(|000\rangle+|111\rangle)
$$
となり、これが3量子ビットのGHZ状態である。測定確率の理論的な導出方法と実際の測定確率の確認方法は前問と同じである。

## 参考資料
- Qiskitチュートリアル https://qiskit.org/documentation/locale/ja_JP/tutorials/circuits/01_circuit_basics.html#%E5%9B%9E%E7%B7%9A%E3%81%AE%E6%A7%8B%E7%AF%89